<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [3]:
import pandas as pd
import psycopg2

In [5]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''select 
                COUNT(id)
                from public.vacancies
'''

In [1]:
# результат запроса
df = pd.read_sql_query(query_3_1 , connection)
df


NameError: name 'pd' is not defined

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''select
              COUNT(id)
           from public.EMPLOYERS
'''

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_12680\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''select
              COUNT(id)
           from public.AREAS
'''

In [9]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_12680\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''select
              COUNT(id)
           from public.INDUSTRIES
'''

In [11]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_12680\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

# Выводы по предварительному анализу данных
В базе данных присутствует 49197 вакансий, 23501 работодателей, 1362 регионов и 294 сферы деятельности. Эти данные связаны между собой по набору id.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_4_1 = f'''SELECT 
    a.name as areas_name,
    COUNT(v.id) as vacancies
FROM   
    public.areas a
JOIN public.VACANCIES v ON v.area_id = a.id
GROUP BY areas_name
ORDER BY 2 DESC
'''

In [13]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_12680\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


areas_name  vacancies
0             Москва       5333
1    Санкт-Петербург       2851
2              Минск       2112
3        Новосибирск       2006
4             Алматы       1892
..               ...        ...
764       Тарко-Сале          1
765    Новоаннинский          1
766            Бирск          1
767           Сасово          1
768             Анна          1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = f'''SELECT 
    COUNT(VACANCIES.id) as "Количество вакансий"
FROM   
    public.VACANCIES
    WHERE salary_from IS NOT NULL or salary_to IS NOT NULL
'''

In [15]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_12680\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


Количество вакансий
0                24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3 = f'''SELECT 
    round (AVG(salary_from)) AS "Среднее значение по нижней границе заработной платы",
    round (AVG(salary_to)) AS "Среднее значение по верхней границе заработной платы"
FROM   
    public.VACANCIES
'''

In [17]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_12680\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


Среднее значение по нижней границ  Среднее значение по верхней грани
0                            71065.0                           110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''SELECT 
    VACANCIES.schedule as "Тип рабочего графика",
    VACANCIES.employment as "Тип трудоустройства",
    COUNT(VACANCIES.id) as "Вакансии"
FROM   
    public.VACANCIES
GROUP BY VACANCIES.schedule,VACANCIES.employment
ORDER BY 3 DESC
'''

In [19]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_12680\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


Тип рабочего графика  Тип трудоустройства  Вакансии
0           Полный день     Полная занятость     35367
1      Удаленная работа     Полная занятость      7802
2         Гибкий график     Полная занятость      1593
3      Удаленная работа  Частичная занятость      1312
4        Сменный график     Полная занятость       940
5           Полный день           Стажировка       569
6        Вахтовый метод     Полная занятость       367
7           Полный день  Частичная занятость       347
8         Гибкий график  Частичная занятость       312
9           Полный день     Проектная работа       141
10     Удаленная работа     Проектная работа       133
11        Гибкий график           Стажировка       116
12       Сменный график  Частичная занятость       101
13     Удаленная работа           Стажировка        64
14        Гибкий график     Проектная работа        18
15       Сменный график           Стажировка        12
16       Вахтовый метод     Проектная работа         2
17       Сменный график     Проектная работа         1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
query_4_5 = f'''SELECT 
    VACANCIES.experience as "Опыт работы",
    COUNT(VACANCIES.id) as "Количество вакансий"
FROM   
    public.VACANCIES
GROUP BY VACANCIES.experience
ORDER BY 2 ASC
'''

In [21]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_12680\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


Опыт работы  Количество вакансий
0         Более 6 лет                 1337
1           Нет опыта                 7197
2       От 3 до 6 лет                14511
3  От 1 года до 3 лет                26152

***

# Выводы по детальному анализу вакансий
Из выполненных запросов можно сделать вывод, что в базе данных на большие города приходится больше вакансий, чем на маленькие города/регионы. Заработная плата находится в диапазоне примерно от 71 тыс. рублей до 110 тыс. рублей (причем хотя бы одно поле с зарплатой заполнено только в половине вакансий). Также заметно, что велик спрос на трудоустройство по типу "Полная занятость" вне зависимости от типа рабочего графика. Кроме того было выявлено, что среди представленных вакансий больше всего соискателей с опытом работы от 1 года до 3 лет, а меньше всего - с опытом работы более 6 лет.


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
query_5_1 = f'''(SELECT 
    EMPLOYERS.name as "Работодатель",
    COUNT(VACANCIES.id) as "Количество вакансий"
FROM   
    public.EMPLOYERS
JOIN public.VACANCIES ON VACANCIES.employer_id = EMPLOYERS.id
GROUP BY EMPLOYERS.name
ORDER BY 2 DESC
LIMIT 1)
UNION ALL
(SELECT 
    EMPLOYERS.name as "Работодатель",
    COUNT(VACANCIES.id) as "Количество вакансий"
FROM   
    public.EMPLOYERS
JOIN public.VACANCIES ON VACANCIES.employer_id = EMPLOYERS.id
GROUP BY EMPLOYERS.name
ORDER BY 2 DESC
OFFSET 4 LIMIT 1)

'''

In [23]:
# Дополнительный запрос: вывод работодателей по убыванию количества вакансий
query_5_1_extra = f'''SELECT 
    EMPLOYERS.name as "Работодатель",
    COUNT(VACANCIES.id) as "Количество вакансий"
FROM   
    public.EMPLOYERS
JOIN public.VACANCIES ON VACANCIES.employer_id = EMPLOYERS.id
GROUP BY EMPLOYERS.name
ORDER BY 2 DESC
'''

In [24]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_12680\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


Работодатель  Количество вакансий
0         Яндекс                 1933
1  Газпром нефть                  331

In [25]:
# результат дополнительного запроса
df = pd.read_sql_query(query_5_1_extra, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_12680\1705873245.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1_extra, connection)


Работодатель  Количество вакансий
0                      Яндекс                 1933
1                  Ростелеком                  491
2                    Тинькофф                  444
3                        СБЕР                  428
4               Газпром нефть                  331
...                       ...                  ...
14761            им. Тьюринга                    1
14762        LLC Novalab Tech                    1
14763   Компания Мир упаковки                    1
14764  Лифинцев Олег Иванович                    1
14765      ТрансЭнергоХолдинг                    1

[14766 rows x 2 columns]

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [26]:
# текст запроса
query_5_2 = f'''SELECT 
    a.name as "Название региона",
    COUNT(distinct e.id) as "Количество работодателей",
    COUNT(distinct v.id) as "Количество вакансий"
FROM   
    public.AREAS a
JOIN public.EMPLOYERS e ON e.area = a.id
LEFT JOIN public.VACANCIES v ON v.area_id = a.id
WHERE v.id is NULL
GROUP BY a.name
ORDER BY 2 DESC

'''

In [27]:
# Дополнительный запрос: список работодателей и регионов по убыванию количества вакансий
query_5_2_extra = f'''SELECT 
    a.name as "Название региона",
    COUNT( distinct e.id) as "Количество работодателей",
    COUNT(distinct v.id) as "Количество вакансий"
FROM   
    public.AREAS a
LEFT JOIN public.EMPLOYERS e ON a.id = e.area
LEFT JOIN public.VACANCIES v ON v.area_id = a.id

GROUP BY "Название региона"
ORDER BY  "Количество работодателей" desc, "Количество вакансий" DESC


'''

In [28]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_12680\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


Название региона  Количество работодателей  \
0                                       Россия                       410   
1                                    Казахстан                       207   
2                           Московская область                        75   
3                           Краснодарский край                        19   
4                                     Беларусь                        18   
..                                         ...                       ...   
181                                  Миллерово                         1   
182  Михайловка (Уфимский район, Башкортостан)                         1   
183                                    Можайск                         1   
184                                Морозовск\n                         1   
185                                     Несвиж                         1   

     Количество вакансий  
0                      0  
1                      0  
2                      0  
3                      0  
4                      0  
..                   ...  
181                    0  
182                    0  
183                    0  
184                    0  
185                    0  

[186 rows x 3 columns]

In [29]:
# результат дополнительного запроса
df = pd.read_sql_query(query_5_2_extra, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_12680\600172942.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2_extra, connection)


Данный запрос нерезультативен, т.к. на его выполнение требуется слишком много времени, поэтому стоит использовать более конкретные запросы вместо него.

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [ ]:
# текст запроса
query_5_3 = f'''SELECT 
    e.name as "Работодатель",
    COUNT(distinct v.area_id) as "Количество регионов, где публикуются вакансии",
    COUNT(distinct v.id) as "Количество вакансий"
FROM   
    public.EMPLOYERS e
JOIN public.VACANCIES v ON v.employer_id = e.id
GROUP BY e.name
ORDER BY 3 DESC
'''

In [ ]:
# результат запроса
df = pd.read_sql_query (query_5_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_5052\2827807884.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_5_3, connection)


Работодатель  \
0                                     Яндекс   
1                                 Ростелеком   
2                                   Тинькофф   
3                                       СБЕР   
4                              Газпром нефть   
...                                      ...   
14761             ПСБ Инновации и Инвестиции   
14762                             ПСЗ Янтарь   
14763                                    ПСК   
14764  Псковский завод механических приводов   
14765                     ПСК ТатЭнергоСтрой   

       Количество регионов, где публикую  Количество вакансий  
0                                    181                 1933  
1                                    152                  491  
2                                     43                  444  
3                                     24                  428  
4                                     24                  331  
...                                  ...                  ...  
14761                                  1                    1  
14762                                  1                    1  
14763                                  1                    1  
14764                                  1                    1  
14765                                  1                    1  

[14766 rows x 3 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [ ]:
# текст запроса
query_5_4 = f'''SELECT
    COUNT(distinct e.id) as "Количество работодателей"
FROM   
    public.EMPLOYERS e
LEFT JOIN public.EMPLOYERS_INDUSTRIES i ON e.id = i.employer_id
WHERE i.industry_id IS NULL
'''

In [ ]:
# результат запроса
df = pd.read_sql_query (query_5_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_5052\796163143.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_5_4, connection)


Количество работодателей
0                      8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [ ]:
# текст запроса
query_5_5 = f'''SELECT
    e.name as "Название работодателя",
    COUNT(distinct i.industry_id) as "Количество сфер деятельности"
FROM   
    public.EMPLOYERS e
LEFT JOIN public.EMPLOYERS_INDUSTRIES i ON e.id = i.employer_id
GROUP BY e.name
HAVING COUNT(distinct i.industry_id) = 4
ORDER BY 1 ASC
OFFSET 2 LIMIT 1
'''

In [ ]:
# Дополнительный запрос: список работодателей по убыванию количества сфер деятельности
query_5_5_extra = f'''SELECT
    e.name as "Название работодателя",
    COUNT(distinct i.industry_id) as "Количество сфер деятельности"
FROM   
    public.EMPLOYERS e
LEFT JOIN public.EMPLOYERS_INDUSTRIES i ON e.id = i.employer_id
GROUP BY e.name

ORDER BY 2 desc

'''

In [ ]:
# результат запроса
df = pd.read_sql_query (query_5_5, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_5052\3433731835.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_5_5, connection)


Название работодателя  Количество сфер деятельности
0                  2ГИС                             4

In [ ]:
# результат запроса
df = pd.read_sql_query (query_5_5_extra, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_13980\2375484581.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_5_5_extra, connection)


Название работодателя  \
0                                               Модуль   
1                                              Энергия   
2                                                Прайд   
3                                                Альфа   
4                                                  СБК   
...                                                ...   
23170         Ярославский завод упаковочных материалов   
23171                         Ярославский филиал ПГУПС   
23172  Ярославский ЦГМС - филиал ФГБУ Центральное УГМС   
23173                                     Ярпож Казань   
23174                          Яшнова Алена Викторовна   

       Количество сфер деятельности  
0                                16  
1                                10  
2                                10  
3                                 9  
4                                 9  
...                             ...  
23170                             0  
23171                             0  
23172                             0  
23173                             0  
23174                             0  

[23175 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [ ]:
# текст запроса
query_5_6 = f'''SELECT
    COUNT(distinct e.id) as "Количество работодателей"
FROM   
    public.EMPLOYERS e   
LEFT JOIN public.EMPLOYERS_INDUSTRIES i ON e.id = i.employer_id
JOIN public.INDUSTRIES y ON y.id = i.industry_id 
WHERE y.name = 'Разработка программного обеспечения'
'''

In [ ]:
# результат запроса
df = pd.read_sql_query (query_5_6, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_5052\1053093668.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_5_6, connection)


Количество работодателей
0                      3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [ ]:
# код для получения списка городов-милионников
from urllib.parse import quote_plus
url = quote_plus('https://ru.wikipedia.org/wiki/Города-миллионеры_России', safe=':/')
tables = pd.read_html(url)
mln_cities = tables[1]['Город'].values.tolist()

In [ ]:
# текст запроса
query_5_7 = f'''
    select
        a.name as region,
        count(*) as vacancies_count
    from
        public.vacancies as v
        join public.employers as e on v.employer_id = e.id
        join public.areas as a on v.area_id = a.id
    group by
        v.area_id, a.name, e.id, e.name
    having
        a.name in {tuple(mln_cities)} and e.name = 'Яндекс'
    

    union all
    

    select
        'Total',
        sum(x.cnt)
    from
        (
        select
            count(*) as cnt
        from
            public.vacancies as v
            join public.employers as e on v.employer_id = e.id
            join public.areas as a on v.area_id = a.id
        group by
            v.area_id, a.name, e.id, e.name
        having
            a.name in {tuple(mln_cities)} and e.name = 'Яндекс'
        ) as x
        

    order by 2
    '''


In [ ]:
# результат запроса
df = pd.read_sql_query (query_5_7, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_5052\810614778.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_5_7, connection)


region  vacancies_count
0              Омск             21.0
1         Челябинск             22.0
2        Красноярск             23.0
3         Волгоград             24.0
4    Ростов-на-Дону             25.0
5            Казань             25.0
6             Пермь             25.0
7               Уфа             26.0
8            Самара             26.0
9         Краснодар             30.0
10          Воронеж             32.0
11      Новосибирск             35.0
12  Нижний Новгород             36.0
13     Екатеринбург             39.0
14  Санкт-Петербург             42.0
15           Москва             54.0
16            Total            485.0

***

# Выводы по анализу работодателей
C помощью запросов в данном юните было выявлено, какие работодатели предоставляют наибольшее количество вакансий (в топ-3 входят компании Яндекс, Ростелеком и Тинькофф), также увидеть, в каком регионе сколько вакансий доступно от какого количества работодателей (топ-3 региона ...). Кроме того, с помощью похожего запроса можно определить, что Яндекс, Ростелеком и Тинькофф охватывают наибольшее количество регионов по количеству вакансий, доступных в них. К сожалению, можно заметить, что больше 8 тыс. работодателей не указывают в вакансии свою сферу деятельности. Но среди тех, кто сферы деятельности указывает, по их количеству лидируют компании Модуль, Энергия и Прайд. Также можно получить данные о работодателях, размещающих вакансии в конкретных сферах (был разобран пример сферы разработки программного обеспечения) и даже получить информацию о вакансиях по конкретным компаниям (например, на Москву приходится 54 вакансии в компании Яндекс).

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
# текст запроса
query_6_1 = '''select count(*)
                from public.vacancies
                where name ilike '%data%' or name ilike '%данн%'-- ilike не учитывает регистр
'''

In [ ]:
# результат запроса
df = pd.read_sql_query (query_6_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_5052\2916403832.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
# текст запроса
query_6_2 = '''select count(*)
    from
        public.vacancies v
    where -- условия к названиям вакансий
        (v.name ILIKE '%data scientist%' OR
        v.name ILIKE '%data science%'OR
        v.name ILIKE '%исследователь данных%' OR
        (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
        v.name ILIKE '%machine learning%' OR v.name ILIKE '%машинн%обучен%')
        and -- условия для Junior-вакансий
        (v.name ILIKE '%junior%' or
        v.experience = 'Нет опыта' or
        v.employment ILIKE '%стажировка%')
'''

In [ ]:
# результат запроса
df = pd.read_sql_query (query_6_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_13980\744361491.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# текст запроса
query_6_3 = '''select count(*)
    from
        public.vacancies v
    where
        (v.name ILIKE '%data scientist%' OR
        v.name ILIKE '%data science%' OR
        v.name ILIKE '%исследователь данных%' OR
        (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
        v.name ILIKE '%machine learning%' OR v.name ILIKE '%машинн%обучен%')
        and -- условия для SQL и postgres
        (v.key_skills ILIKE '%SQL%' or
        v.key_skills LIKE '%postgres%')
'''

In [ ]:
# результат запроса
df = pd.read_sql_query (query_6_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_5052\2436245350.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# текст запроса
query_6_4 = '''select count(*)
    from
        public.vacancies v
    where
        (v.name ILIKE '%data scientist%' OR
        v.name ILIKE '%data science%' OR
        v.name ILIKE '%исследователь данных%' OR
        (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
        v.name ILIKE '%machine learning%' OR v.name ILIKE '%машинн%обучен%')
        and -- наличие ключевого навыка Python
        (v.key_skills ILIKE '%Python%' )
'''

In [ ]:
# результат запроса
df = pd.read_sql_query (query_6_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_5052\4166993872.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
# текст запроса
query_6_5 = '''select 
            round(avg(length(key_skills)-length(replace(key_skills, CHR(9), ''))+1), 2) as result
    from
        public.vacancies v
    where
        (v.name ILIKE '%data scientist%' OR
        v.name ILIKE '%data science%' OR
        v.name ILIKE '%исследователь данных%' OR
        (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
        v.name ILIKE '%machine learning%' OR v.name ILIKE '%машинн%обучен%')
        
'''

In [ ]:
# результат запроса
df = pd.read_sql_query (query_6_5, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_5052\1144307522.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_6_5, connection)


result
0    6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
# текст запроса
query_6_6 = '''select 
            v.experience as "Опыт работы",
            ROUND(AVG(COALESCE((salary_from + salary_to) / 2, salary_from, salary_to)), 0) as "Средний размер зарплаты"
    from
        public.vacancies v
    where
        (v.name ILIKE '%data scientist%' OR
        v.name ILIKE '%data science%' OR
        v.name ILIKE '%исследователь данных%' OR
        (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
        v.name ILIKE '%machine learning%' OR v.name ILIKE '%машинн%обучен%')
        AND v.experience = 'От 3 до 6 лет'
        AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
        GROUP BY v.experience
        
'''

In [ ]:
# результат запроса
df = pd.read_sql_query (query_6_6, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_5052\3310421505.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query (query_6_6, connection)


Опыт работы  Средний размер зарплаты
0  От 3 до 6 лет                 243115.0

***

# Выводы по предметному анализу
Благодаря предметному анализу можно получить нужную информацию из текстовых данных. На примере вакансий, связанных с Data Science, было получено, что к работе с данными относятся более 1,7 тыс. вакансий,51 из которых подходят специалистам уровня Junior. Также можно проверить наличие в вакансиях конкретных ключевых навыков (например, 201 вакансия относится к SQL или postgres), узнать количество требуемых ключевых навыков (в среднем их 6,41 для DS). Кроме того, можно увидет среднюю зарплату для специалиста с определенным количеством опыта (например, больше 240 тыс. рублей для специалиста с опытом 3-6 лет работы).

In [6]:
connection.close()

# Общий вывод по проекту

По итогу проведенного анализа можно сделать вывод, что предоставленная база данных имеет обширную информацию о вакансиях, предоставляющих их работодателях, регионах, в которых они представлены. Ожидаемо, наибольшее количество вакансий предоставляется в крупных городах большими компаниями. Также было выявлено, кандидатов с каким опытом работы и на какой тип занятости требуется больше всего, а также какую зарплату работодатели в среднем готовы обеспечивать. Кроме того, из данного датасета можно также извлечь информацию, касающююся отдельных отраслей, что было рассмотрено на примере DS. Будет полезно провести дальнейшие исследования на предмет обнаружения самых актуальных и востребованных сфер занятости, а также навыков и опыта, требуемых для них.